inception-active-learning/src/main/java/de/tudarmstadt/ukp/inception/active/learning/ActiveLearningServiceImpl.java

```diff
@@ -102,4 +114,339 @@ public ActiveLearningServiceImpl(DocumentService aDocumentService,
+    {
+        ActiveLearningUserState state = getState(aUser.getUsername(), aProject);
+        AnnotationLayer selectedLayer;
+        synchronized (state) {
+            selectedLayer = state.getSelectedLayer(aProject);
+        }
+        return selectedLayer;
+    }
+
+    @Override
+    public void putActiveLearningRecommender(User aUser, Project aProject,
+        ActiveLearningRecommender aActiveLearningRecommender)
+    {
+        ActiveLearningUserState state = getState(aUser.getUsername(), aProject);
+        synchronized (state) {
+            state.setActiveLearningRecommender(aProject, aActiveLearningRecommender);
+        }
+    }
+
+    @Override
+    public ActiveLearningRecommender getActiveLearningRecommender(User aUser, Project aProject)
+    {
+        ActiveLearningUserState state = getState(aUser.getUsername(), aProject);
+        ActiveLearningRecommender activeLearningRecommender;
+        synchronized (state) {
+            activeLearningRecommender = state.getActiveLearningRecommender(aProject);
+        }
+        return activeLearningRecommender;
+    }
+
+    @Override
+    public void putLearnSkippedRecommendationTime(User aUser, Project aProject,
+        Date aLearnSkippedRecommendationTime)
+    {
+        ActiveLearningUserState state = getState(aUser.getUsername(), aProject);
+        synchronized (state) {
+            state.setLearnSkippedRecommendationTime(aProject, aLearnSkippedRecommendationTime);
+        }
+    }
+
+    @Override
+    public Date getLearnSkippedRecommendationTime(User aUser, Project aProject)
+    {
+        ActiveLearningUserState state = getState(aUser.getUsername(), aProject);
+        Date learnSkippedRecommendationTime;
+        synchronized (state) {
+            learnSkippedRecommendationTime = state.getLearnSkippedRecommendationTime(aProject);
+        }
+        return learnSkippedRecommendationTime;
+    }
+
+    private ActiveLearningUserState getState(String aUsername, Project aProject)
+    {
+        synchronized (states) {
+            ActiveLearningUserState state;
+            state = states.get(aUsername);
+            if (state == null) {
+                state = new ActiveLearningUserState();
+                state.initializeActiveLearningUserState(aProject);
+                states.put(aUsername, state);
+            }
+            return state;
+        }
+    }
+
+    @EventListener
+    public void onActiveLearningSessionCompleted(ActiveLearningSessionCompletedEvent aEvent)
+    {
+        clearState(aEvent.getUser());
+    }
+
+    @EventListener
+    public void onDocumentOpened(DocumentOpenedEvent aEvent)
+    {
+        clearState(aEvent.getUser());
+    }
+
+    private void clearState(String aUserName)
+    {
+        synchronized (states) {
+            states.remove(aUserName);
+        }
+    }
+
+    private static class ActiveLearningUserState
+    {
+        private Map<Long, Boolean> sessionActive = new HashMap<>();
```

Comment:
Instead of having a map for every single variable, why not have a higher-level map with the project ID as the key and the `ActiveLearningUserState` as the value?